---
title: "Chromosomal Aberration Analysis for UHCW"
author: "JING YUAN"
date: "9/13/2018"
output: 
   prettydoc::html_pretty:
    theme: cayman
    highlight: github
    toc: TRUE
    
---
#Analysis of genomic integrity using eSNP-Karyotyping
#####Glioblastomas is characterized as heterogeneous disease, and the effect of chromosomal aberrations have been commonly found in the tumorogenesis. According to the previous study, we found genes relevant to centromere are highly expressed in our dataset. Here we adopted eSNP-Karyotyping for our RNA-Seq dataset and present eSNP-Karyotyping for the detection of chromosomal aberrations, based on measuring the ratio of expression between the two alleles.

#####The analysis pipeline is separated into duplications and LOHs detection. The output of the duplication part of the analysis is a moving median graph of the allelic ratios sorted by the chromosomal position.The statistical test is calculated with a one tailed t-test comparing the SNPs major/minor values in each window to the values of the entire genome. The p-values are FDR corrected and displayed as color bar above the moving median plot.

#####The Loss of heterozygosity (LOH) analysis is performed by merging the detected SNPs table with a table of the common SNPs, and comparing the distributions of the homozygous and heterozygous positions. Large stretches of homozygosity are considered regions with LOH. The statistical test is performed at the chromosomal arm scale. To obtain p-value, the ratio of the number of homozygote to heterozygote SNPs for each chromosome arm is calculated. Then, arms with statistically significant different ratio are recognized by t-test. The p-values are adjusted for multiple testing using FDR correction. True LOH is considered as an arm with p-value < 0.001 and a homozygote to heterozygote SNPs ratio five times greater than the ratio of all the autosomal chromosomes. Arm with significant difference in the distribution of the homozygous to heterozygous SNPs will be highlighted in yellow.

##Scripts done in Linux in sever
###1. Trim the sequences
```{bash}
for i in WTCHG_109791_02 WTCHG_109791_04 WTCHG_109791_05 WTCHG_109791_06 WTCHG_110084_07 WTCHG_110084_12 WTCHG_110084_13 WTCHG_110084_14 WTCHG_110734_15 WTCHG_110734_16 WTCHG_110734_18 WTCHG_110734_19;  do java -jar ~/programms/Trimmomatic-0.33/trimmomatic-0.33.jar PE -phred33 ${i}_1.fastq ${i}_2.fastq  ../3_trimmed/${i}_1.fastq.gz ../3_trimmed/${i}_1UP.fastq.gz ../3_trimmed/${i}_2.fastq.gz ../3_trimmed/${i}_2UP.fastq.gz ILLUMINACLIP:../../programms/Trimmomatic-0.33/adapters/TruSeq3-PE.fa:2:30:10 LEADING:24 TRAILING:24 SLIDINGWINDOW:4:20 MINLEN:40 ; done
```

###2. Prepare genome indexes 
```{bash}
bowtie2-build /home/u1566273/reference_genome/hg38/hg38.fa /home/u1566273/reference_genome/hg38/hg38
```

###3. Build the SAMtools genome index 
```{bash}
samtools faidx /home/u1566273/reference_genome/hg38/hg38.fa
```

###4.Build the Picard genome dictionary file run
```{bash}
java -jar /home/u1566273/program/picard.jar CreateSequenceDictionary R=/home/u1566273/reference_genome/hg38/hg38.fa O=/home/u1566273/reference_genome/hg38/hg38.dict
```



##Scripts done in R programme in Linux in sever
###5. Load libraries
```{bash}
install_github("BenvenLab/eSNPKaryotyping/eSNPKaryotyping")
library("devtools")
library("zoo")
library("gplots")
library("eSNPKaryotyping")
```


###6. The downloaded dbSNP database have to be restructured 
```{bash}
Edit_dbSNP_Files(Directory= "/home/u1566273/reference_genome/hg38/dbsnp_hg38/", File_Name="", Organism= "Human")
```

###7. Alignment of the reads to the reference
```{bash}
Tophat(Directory = "/home/u1566273/esnp/FASTQ/”, Library_Type = "Paired", Threads = 8, Transcripts_Annotation = “/home/u1566273/reference_genome/hg38/hg38.gtf”, Bowtie_Genome_Index = “/home/u1566273/reference_genome/hg38/hg38” ) 
```

###8. Calling SNPs from the align BAM file 
```{bash}
CreateVCF(Directory= "~/esnp/BAM/“, Genome_Fa= "~/reference_genome/hg38/hg38.fa", Picard_Path= "~/program/", GATK_Path= "~/program/")
```

###9. Add read groups information 
```{bash}
#make sure work in the right directory
java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam  O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_110734_19 LB= lib12 PL=illumina PU=lane6 PM=Hiseq RGSM=K8T3

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_110084_13 LB= lib7 PL=illumina PU=lane5 PM=Hiseq RGSM=K8N1

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_109791_02 LB= lib1 PL=illumina PU=lane4 PM=Hiseq RGSM=K2N1

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_109791_04 LB= lib2 PL=illumina PU=lane4 PM=Hiseq RGSM=K2HG3

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_109791_05 LB= lib3 PL=illumina PU=lane4 PM=Hiseq RGSM=K3HG3

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_109791_06 LB= lib4 PL=illumina PU=lane4 PM=Hiseq RGSM=K4HG3

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_110084_07 LB= lib5 PL=illumina PU=lane5 PM=Hiseq RGSM=K4N1

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_110084_12 LB= lib6 PL=illumina PU=lane5 PM=Hiseq RGSM=K6HG3

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_110084_14 LB= lib8 PL=illumina PU=lane5 PM=Hiseq RGSM=K9N01

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_110734_15 LB= lib9 PL=illumina PU=lane6 PM=Hiseq RGSM=K9HG3

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_110734_16 LB= lib10 PL=illumina PU=lane6 PM=Hiseq RGSM=K3N2

java -jar /home/u1566273/program/picard.jar AddOrReplaceReadGroups I=accepted_hits.bam O=accepted_hits_rg.bam SO=coordinate ID=WTCHG_110734_18 LB= lib11 PL=illumina PU=lane6 PM=Hiseq RGSM=K6N1

#check the read groups information
samtools view -H accepted_hits_rg.bam | grep '@RG'
```

###10. Reorder bam files, create index and mark duplicates
```{bash}
java -jar /home/u1566273/program/picard.jar ReorderSam REFERENCE=/home/u1566273/reference_genome/hg38/hg38.fa INPUT=accepted_hits_rg.bam OUTPUT=accepted_hits_rg_sorted.bam

java -jar /home/u1566273/program/picard.jar BuildBamIndex I=accepted_hits_rg_sorted.bam

java -jar /home/u1566273/program/picard.jar MarkDuplicates I=accepted_hits_rg_sorted.bam O=dedupped.bam CREATE_INDEX=true VALIDATION_STRINGENCY=SILENT M=output.metrics
 
```

###11. Processing and filtering the variants table
```{bash}
#This function creates file with SNPs data at the BAM directory called variantTable.csv 
table=EditVCF(Directory=“~/esnp/BAM/“, Organism=“Human”)
```

###12. Sort the SNP table according to the chromosomal position
```{r}
table= read.table("/Volumes/eSNP/K2N/BAM/variantTable.csv",header = TRUE)
table$chr=as.numeric(table$chr)
table=table[order(table$chr,table$position),]
table=table[table$chr>0,]
```

###13. Filter the SNPs according to specified parameters and calculates the allelic ratio of each SNP 
```{r}
table2=MajorMinorCalc(Table=table, minDP=20, maxDP=10000, minAF=0.2)

```

###14. Plot Allelic ratio along the genome for duplication detection:
```{r}
plot_true=PlotGenome(orderedTable=table2, Window=151, Ylim=3, PValue="TRUE", Organism="Human")

```

###15. Intersects the observed SNPs with the list of the common SNPs from dbSNP database
```{r}
tbl=DeletionTable(Directory="/Volumes/eSNP/K2N/BAM/", Table=table2, dbSNP_Data_Directory="/Volumes/eSNP/reference/hg38/dbsnp_hg38/", dbSNP_File_Name= "Edited_Common_chr", Genome_Fa_dict = "/Volumes/eSNP/reference/hg38/hg38.dict", Organism="Human")

```

###16. Plot each SNP 
```{r}
Plot_Zygosity_Sinle(Table =tbl, Organism = "Human")
```

###17. Plot blocks of heterozygous and homozygous SNPs
```{r}
Plot_Zygosity_Blocks(Table = tbl, Window = 1500000, Max = 6, Max2 = 60, Organism = "Human")

```


