---
title: "#RNA-seq analysis of PRRX1 overxpressed Hek293 cell lines"
author: "JING YUAN"
date: "25/11/2021"
---
# Log into vettel
```{bash}
sftp://vettel.bioinf.warwick.ac.uk
ssh -X u1566273@vettel.bioinf.warwick.ac.uk
```

# Mount share folder
```{bash}
mount ~/mnt/Shared210
```

# Full path of fast files
```{bash}
/home/u1566273/mnt/Shared210/Read_database/Human/2021_11_22_Jing_cell_lines_RNA-seq
```

# Update program on vettel
```{bash}
conda create --name <name of environment you want to create>

# When asked if you want to install [y]/n type y and press enter

conda activate <name of environment you created>

# Google package you want to install and copy the first install command e.g.

conda install -c bioconda fastqc

# When asked if you want to install [y]/n type y and press enter
# When installed run:

conda activate <name of environment you created>

# Run the command you want to run

# When done type to exit the environment and return to normal vettel
 
conda deactivate

▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀
▓▓   ▓▓ Install new version of SAMtools     ▓▓
▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀
samtools # unupdated version
mkdir ~/bin #  inside ~/.profile there's a line adding ~/bin to $PATH
mkdir ~/programs

cd ~/programs 
# get link from htslib.org/download/ (right click, copy link adress)
wget https://github.com/samtools/samtools/releases/download/1.14/samtools-1.14.tar.bz2
tar -jxf samtools-1.*.tar.bz2 # unzip
cd samtools-1.*/
./configure --prefix=$HOME # tell where to install samtools (that/bin must be in the $PATH)
make 
make install

```

# First quality check
```{bash}
conda create --name fastqc
echo ". /shared/software/miniconda3/etc/profile.d/conda.sh" >> ~/.bashrc
source ./.bashrc
conda install -c bioconda fastqc
conda activate fastqc
conda deactivate fastqc
fastqc -o /home/u1566273/RNASEQ_data/1_fastqc/ *.fq.gz
multiqc .
```

# Quality control and trimming with Fastp
```{bash}
#change file names
for i in *_1.fq.gz; do mv -- "$i" "${i%_1.fq.gz}.R1.fq.gz"& done
for i in *_2.fq.gz; do mv -- "$i" "${i%_1.fq.gz}.R2.fq.gz"& done

#trimming fastq.gz files
conda activate fastp

for i in *.R1.fq.gz; do fastp --thread 16 --length_required 40 --cut_right --cut_window_size 4 --cut_mean_quality 20 --overrepresentation_analysis -i $i -I ${i/.R1./.R2.} -o /home/u1566273/mnt/Shared342/Jing/2_trimmed/${i/.R1./.R1.trimmed.} -O /home/u1566273/mnt/Shared342/Jing/2_trimmed/${i/.R1./.R2.trimmed.} -h /home/u1566273/mnt/Shared342/Jing/2_trimmed/${i%.R1.fq.gz}_fastp.html -j /home/u1566273/mnt/Shared342/Jing/2_trimmed/${i%.R1.fq.gz}_fastp.json & done
```

# Multiqc reports of fastp
```{bash}
cd $dir
multiqc --filename /home/u1566273/RNASEQ_data/2_trimmed/qa_fastp_multiqc /home/u1566273/RNASEQ_data/2_trimmed/

fastqc -o /home/u1566273/RNASEQ_data/3_fastqc/ *.fq.gz
```

# STAR Alignment
## Build the STAR index
```{bash}
#update star on vettel
cd ~/programs # Get latest STAR source from releases
wget https://github.com/alexdobin/STAR/archive/2.7.9a.tar.gz
tar -xzf 2.7.9a.tar.gz

# Compile
cd STAR-2.7.9a
cd source
make STAR

# Add to the PATH
echo 'export PATH=$HOME/program/STAR-2.7.9a/source:$PATH' >> ~/.profile
source ~/.profile

STAR --version

cd $dir
genome="$/home/u1566273/reference_genome/human_genome/GRCh38.p13.genome.fa.gz" # double quotes important
annotation="$/home/u1566273/reference_genome/human_genome/gencode.v38.annotation.gtf"
gunzip gencode.v38.annotation.gtf GRCh38.p13.genome.fa.gz 
STAR --runThreadN 20 --runMode genomeGenerate --genomeDir /home/u1566273/reference_genome/human_genome --genomeFastaFiles /home/u1566273/reference_genome/human_genome/GRCh38.p13.genome.fa --sjdbGTFfile /home/u1566273/reference_genome/human_genome/gencode.v38.annotation.gtf --sjdbOverhang 149

```

## Mapping
```{bash}
cd /home/u1566273/RNASEQ_data/2_trimmed
time_file="../timing_mapping_STAR.tsv"
printf "sample\tstart\tend\n" > $time_file
for i in *R1.trimmed.fq.gz; do 
  printf "$i\t$(date '+%Y/%m/%d %H:%M:%S')\t" >> $time_file &&
  STAR --runThreadN 10 --genomeDir /home/u1566273/reference_genome/human_genome \
    --readFilesIn $i ${i/.R1./.R2.} --readFilesCommand zcat --outSAMtype BAM SortedByCoordinate \
    --outFileNamePrefix /home/u1566273/mnt/Shared342/Jing/4_alignment/${i/.R1.*/.}  --outWigType bedGraph --outWigStrand Unstranded \
    --quantMode GeneCounts --outTmpDir ../4_alignment/tmp/${i/.R1.*/.} &&
  printf "$(date '+%Y/%m/%d %H:%M:%S')\n" >> $time_file ;
done
```

#umount the share folders
```{bash}
umount mnt/Shared342 #disconnect
mount mnt/Shared342 #connect to share folder
(fastqc) u1566273@vettel:~$ ls mnt/Shared377/
(fastqc) u1566273@vettel:~$ rmdir mnt/Shared377/
(fastqc) u1566273@vettel:~$ ls mnt/Shared342/
(fastqc) u1566273@vettel:~$ du -sh * #check space
```

```{bash}
# move the counts data and split the stats
mkdir human
cd /Volumes/Jing.Y4_2/Hek293/4_alignment/human
for i in *ReadsPerGene.out.tab; do
  tail -n +5 $i > /Volumes/Jing.Y4_2/Hek293/5_counts/human/${i/.ReadsPerGene.out.tab/.counts.tsv} &&
  mv $i /Volumes/Jing.Y4_2/Hek293/qa/mapped_human
done

#extract first two columns of counting files
awk '{print $1,$2}' JY_1.counts.tsv > ../counting_PRRX1/JY_1.count.tsv
for i in *.counts.tsv; do awk '{print $1,$2}' $i > ../counting_PRRX1/$i & done
```